src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,537 @@
+    const std::shared_ptr<AbstractLQPNode>& lqp,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping) {
+  /**
+   * Crawl the `lqp`, including subquery-LQPs, for usages of parameters from `parameter_mapping`
+   */
+
+  auto optimizable = true;
+  auto correlated_predicate_node_count = size_t{0};
+
+  visit_lqp(lqp, [&](const auto& node) {
+    if (!optimizable) {
+      return LQPVisitation::DoNotVisitInputs;
+    }
+
+    auto is_correlated = false;
+
+    for (const auto& expression : node->node_expressions) {
+      visit_expression(expression, [&](const auto& sub_expression) {
+        if (is_correlated || !optimizable) {
+          return ExpressionVisitation::DoNotVisitArguments;
+        }
+
+        if (const auto subquery_expression = std::dynamic_pointer_cast<LQPSubqueryExpression>(sub_expression)) {
+          const auto& [_, count_in_subquery] =
+              assess_correlated_parameter_usage(subquery_expression->lqp, parameter_mapping);
+          if (count_in_subquery > 0) {
+            correlated_predicate_node_count += count_in_subquery;
+            optimizable = false;
+          }
+        }
+
+        if (const auto parameter_expression =
+                std::dynamic_pointer_cast<CorrelatedParameterExpression>(sub_expression)) {
+          if (parameter_mapping.find(parameter_expression->parameter_id) != parameter_mapping.end()) {
+            is_correlated = true;
+          }
+        }
+
+        return is_correlated ? ExpressionVisitation::DoNotVisitArguments : ExpressionVisitation::VisitArguments;
+      });
+    }
+
+    if (is_correlated) {
+      if (node->type == LQPNodeType::Predicate) {
+        ++correlated_predicate_node_count;
+      } else {
+        optimizable = false;
+        return LQPVisitation::DoNotVisitInputs;
+      }
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+
+  return {!optimizable, correlated_predicate_node_count};
+}
+
+std::shared_ptr<BinaryPredicateExpression> SubqueryToJoinRule::try_to_extract_join_predicate(
+    const std::shared_ptr<PredicateNode>& predicate_node,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping, bool is_below_aggregate) {
+  // Check for the type of expression first. Note that we are not concerned with predicates of other forms using
+  // correlated parameters here. We check for parameter usages that prevent optimization later in
+  // assess_correlated_parameter_usage().
+  if (predicate_node->predicate()->type != ExpressionType::Predicate) {
+    return nullptr;
+  }
+
+  const auto& predicate_expression = std::static_pointer_cast<AbstractPredicateExpression>(predicate_node->predicate());
+
+  // We rely on PredicateSplitUpRule having split up ANDed chains of such predicates previously, so that we can process
+  // them separately.
+  auto predicate_condition = predicate_expression->predicate_condition;
+  if (!is_binary_predicate_condition(predicate_condition)) {
```

Comment:
@mrzzzrm This matches more predicate conditions than we want, since it includes `(NOT) LIKE` and `(NOT) IN`. It also causes the sanitizer errors on CI, since we cast an `InExpression` to a `BinaryPredicateExpression`, which it does not derive from

Reply:
Mmpf, we need a function that does what I thought `is_binary_predicate_condition` would do. Will fix this later, sorry for breaking things...